### ISSUE TITLE
Re-Hover doesn't work with 'bottom' settings (in Shopware)

### ISSUE THREAD

**Issue by MLK97** [kamens/jQuery-menu-aim#86]
I use the plugin to display submenu points on the bottom side of the navigation.
It works perfectly, but if I 
1. hover over a point 
2. move mouse downwards until i am outside of the submenu and menu 
3. then go back to the just hovered menu point 

the plugin doesn't execute.

Haven't found a workaround, yet. Maybe it is just because of the framework (Shopware) I use

**Comment by MLK97** [comment]
Problem exists because the activate function gets executed but the deactivate function doesn't get executed

**Comment by rossmedia** [comment]
This bug is a general bug and not related to bottom setting or Shopware.

The bug also exists in the offical demo under:
https://rawgit.com/kamens/jQuery-menu-aim/master/example/example.html

Steps to reproduce:

- [ ]  Open https://rawgit.com/kamens/jQuery-menu-aim/master/example/example.html
- [ ]  Click on "Explode the monkeys" to open sub menu
- [ ]  Mouse over over any of the menu items, let's say "Patas"
- [ ] Move your mouse out to the left and click anywhere on the page so the dropdown will be closed
- [ ] Click again on "Explode the Monkeys" to open sub menu
- [ ] Mouse over over the "Patas" again

**Expected result**
- Drop-Down for "Patas" should open

**Current result**
- Drop-Down does not open

**Comment by stevepiron** [comment]
Quite annoying indeed. I've just found out about this bug, which makes the plugin not a good fit for production.

**Comment by stevepiron** [comment]
I've just found a workaround. It isn't ideal but it works:

```js
/**
 * @function `init`
 *
 * Initializes the desktop navigation behavior.
 *
 * 1. Hide the panel on exit.
 */
function init() {
    $navBar.menuAim({
        rowSelector: navGroup,
        activationDelay: 0,
        activate: function(el) {
            showNavPanel(el);
        },
        deactivate: function(el) {
            hideNavPanel(el);
        },
        exit: function(el) {
            hideNavPanel(el); /* [1] */
        },
        submenuDirection: 'above'
    });
    is_desktop_nav = true;
    enableClickOnPanelLabel();
}

/**
 * When leaving the panel, unbind the mouseenter event
 * and initialize menuAim again. Not ideal as it doesn't destroy
 * the first instance.
 */
$navPanel.on('mouseleave', function(e) {
    e.preventDefault();
    $navGroup.unbind('mouseenter');
    init();
});
```

So basically: hide the nav panel on exit, bind a `mouseleave` event to all panels and initialize menuAim again.

**Comment by MLK97** [comment]
I tried your workaround and it does the job just fine. You basically just double force it. In my own workaround I made it exactly like that, even though it wasn't as elegant as yours.
Hopefully someone will solve this issue

**Comment by zisheng-ai** [comment]
same issue

**Comment by heun01** [comment]
The mouseleaveMenu function only triggers the deactivate function, if the exitMenu option is supplied and returns true:

`
$('.module-nav .module-nav__lvl1').each(function () {
      var $menu = $(this);
      $menu.menuAim({
        activate: activateSubmenu,
        exitMenu: deactivateSubmenu,
        submenuDirection: "below"
      });
    });

function activateSubmenu(row) {
    var $row = $(row);
    $row.addClass('-hover').siblings().removeClass('-hover');
  }

function deactivateSubmenu(row) {
    var $row = $(row);
    $row.find('> .module-nav__item').removeClass('-hover');
    return true;
  }
`

**Comment by stevepiron** [comment]
@heun01 is that how you suggest to fix it? 🤔

**Comment by heun01** [comment]
Yeah works for me.

**Comment by jakedowns** [comment]
I had the same issue and I had overlooked this comment on exitMenu in the example config

if you have a exitMenu option, it needs to return true

```
// Function to call when mouse exits the entire menu. If this returns
// true, the current row's deactivation event and callback function
// will be fired. Otherwise, if this isn't supplied or it returns
// false, the currently activated row will stay activated when the
// mouse leaves the menu entirely.
exitMenu: function (row) {
    return true;
},
```

**Comment by MBWebTechMariusz** [comment]
I've had the same issue, and wanted to share the complete solution:

```
exitMenu: function (row) {
    return true;
},
```

This should be added in the docs, so other people can easily use this in their projects.

**Comment by rossmedia** [comment]
@MBWebTechMariusz Thanks! We had no exitMenu callback implemented so we just added a function like you did and with that it works fine. I think there should be some workaround implemented which fixes that bug and let exitMenu return true by default, even if no callback is implemented.

 


Gesendet: Dienstag, 09. Oktober 2018 um 23:54 Uhr
Von: MBWebTechMariusz <notifications@github.com>
An: kamens/jQuery-menu-aim <jQuery-menu-aim@noreply.github.com>
Cc: rossmedia <d.ross@gmx.org>, Manual <manual@noreply.github.com>
Betreff: Re: [kamens/jQuery-menu-aim] Re-Hover doesn't work with 'bottom' settings (in Shopware) (#86)


I've had the same issue, and wanted to share the complete solution:

exitMenu: function (row) {
    return true;
},


This should be added in the docs, so other people can easily use this in their projects.

**Comment by rossmedia** [comment]
@MBWebTechMariusz Thanks! We had no exitMenu callback implemented so we just added a function like you did and with that it works fine. I think there should be some workaround implemented which fixes that bug and let exitMenu return true by default, even if no callback is implemented.

**Comment by MBWebTechMariusz** [comment]
@rossmedia I think this is not the intended behaviour, see this post: https://github.com/kamens/jQuery-menu-aim/issues/50#issuecomment-68388419
Because of this, I think it is better to make this option as an "opt-in"-option, rather than the default.

What do you think?

**Comment by rossmedia** [comment]
@MBWebTechMariusz Oh yes sorry now i understand the problem. When i implement your solution the benefit of this plugin (work like amazons menu and don't close when you) is not working anymore. :(

**Comment by MBWebTechMariusz** [comment]
Yeah, this is a drawback. But I think this is intended, so we need to try it with just 

```
exitMenu: function (row) {
    return true;
},
```
